Mathlib/Tactic/Eval.lean

```diff
@@ -0,0 +1,44 @@
+/-
+Copyright (c) 2024 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+
+import Qq.Macro
+import Std.Lean.Expr
+
+/-!
+# The `eval%` term elaborator
+
+This file provides the `eval% x` term elaborator, which evaluates the constant `x` at compile-time
+in the interpreter, and interpolates it into the expression.
+-/
+
+
+open Qq Lean Elab Term
+
+namespace Mathlib.Meta
+
+/--
+`eval% x` evaluates the term `x : X` in the interpreter, and then injects the resulting expression.
+
+As an example:
+```lean
+example : 2^10 = eval% 2^10 := by
+  -- goal is `2^10 = 1024`
+  sorry
+```
+This only works if a `Lean.ToExpr X` instance is available.
+-/
+syntax (name := eval_expr) "eval% " term : term
+
+@[term_elab eval_expr, inherit_doc eval_expr]
+unsafe def elabEvalExpr : Lean.Elab.Term.TermElab
+| `(term| eval% $stx) => fun exp => do
+  let e ← Lean.Elab.Term.elabTermAndSynthesize stx exp
+  let e ← instantiateMVars e
+  let ee ← Elab.Term.elabTerm (← `(Lean.toExpr $(← e.toSyntax))) (some q(Expr))
+  Lean.Meta.evalExpr Expr q(Expr) ee (safety := .unsafe)
```

Comment:
```suggestion

The `!`-version `eval%! x` also logs the expression computed by `eval%`.
-/
syntax (name := eval_expr) "eval% " ("!")? term : term

/-- An auxilliary function for converting Syntax to an `Expr`. -/
unsafe def elabEvalExprCore (stx: TSyntax `term) (exp: Option Expr) : TermElabM Expr := do
  let e ← Lean.Elab.Term.elabTermAndSynthesize stx exp
  let e ← instantiateMVars e
  let ee ← Elab.Term.elabTerm (← `(Lean.toExpr $(← e.toSyntax))) (some q(Expr))
  Lean.Meta.evalExpr Expr q(Expr) ee (safety := .unsafe)

@[term_elab eval_expr, inherit_doc eval_expr]
unsafe def elabEvalExpr : Lean.Elab.Term.TermElab
| `(term| eval% $stx) => elabEvalExprCore stx
| `(term| eval%! $stx) => fun exp => do
  let res ← elabEvalExprCore stx exp
  logInfo m!"'eval% {stx}' computed '{← Meta.ppExpr res}'"
  return res
```
What do you think of this?  This way, if you use `eval%` and it does not work, you have the option of seeing what it is that it computed.

For instance, you would get some indication of what failed here:
```lean
#guard_expr
  (eval%! Finset.powerset ({1, 2, 3})) =
    {{1}, {2}, {1, 2}, {3}, {1, 3}, {2, 3}, {1, 2, 3}}
/-
`guard` says the result is not true, but `eval%` says

'eval% Finset.powerset ({1, 2, 3})' computed '{∅, {1}, {2}, {1, 2}, {3}, {1, 3}, {2, 3}, {1, 2, 3}}'
-/
```

Reply:
Isn't this just combining `eval%` with `show_term`?

Reply:
You are probably right, but then maybe it might be useful to mention this in the doc-string, since I had not thought of using `show_term` in this situation.